### **Universidad Central de Venezuela**
### **Facultad de Ingeniería**
### **Escuela de Ingeniería Eléctrica**

## <center>**Laboratorio N°6**<center>

<div style='text-align: right;'>
    <h4>Periodo 2023-3</h4>
    <i>02/2023<br>CI 29571461 Ricardo Santana</i>
</div>

### **1. Introducción**

Cuando se estudian ciertos problemas en ingeniería y física, los modelos matemáticos que surgen corresponden a ecuaciones donde se relaciona la variación de una magnitud con las condiciones internas, externas e iniciales del sistema.

Una ecuación diferencial ordinaria (EDO) expresa la razón de cambio de la variable dependiente y con respecto al cambio de una variable independiente x. La variable dependiente puede ser la población de una bacteria en el tiempo, la temperatura de una sustancia en el tiempo, la concentración de un producto en un reactor. En muchos casos, la variable independiente es el tiempo, ya que se desea medir el cambio de algo (productos, concentración, población, altura, etc.) con respecto al tiempo, dado que cada instante es distinto al anterior.

Para resolver ecuaciones diferenciales en Ingeniería Química, existen muchos métodos que dependen del tipo de ecuación generada en cada problema. De manera general, podemos distinguir dos grandes grupos: los métodos analíticos, que obtienen una solución exacta, pero se ajustan a un solo tipo de ecuación diferencial, y los métodos numéricos, que obtienen una solución aproximada, son más genéricos y dependen de las condiciones iniciales, aunque requieren cálculos repetitivos para corregir y mejorar la aproximación.

Entre los diversos métodos numérico utilizados cabe anlizar el de Euler y Rugen-Kutta, que destacan por su simplicidad y por tanto se verificará que permitan obtener una estimación aproximada a lo que se desea.

### **2. Marco Teórico**

Según [1] la definición Condición de Lipschitz se establece como:

#### **2.1. Definición 1**

Se dice que una función $f (t, y)$ satisface la condición de Lipschitz en la variable y en un
conjunto $D \subset \mathbb{R}^2$ si existe una constante $L > 0$ con

$$|f (t, y_1) - f (t, y_2)| \leq L|y_1 - y_2|$$

#### **2.2. Teorema 1 [1]**

Suponga que  $f (t, y)$ se define sobre un conjunto convexo $D \subset \mathbb{R}^2$. Si existe una constante $L > 0$ con

$$\left| \frac{\partial f}{\partial y}(t,y) \right| \leq L, \;\;\; \forall (t,y) \in D$$

entonces $f$ satisface la condición de Lipschitz en $D$ en la variable $y$ con constante $L$ de Lipschitz

**Demostración:**

Se sabe del teorema del valor medio que,

Si $f \in C[a, b]$ y $f$ es diferenciable en $(a,b)$, entonces existe un número c en $(a,b)$ con

$$f'(c) = \frac{f(b) - f(a)}{b - a}$$

Tomando como valor extremo $f'(c) = L$ y fijamos $t$ en una función $f(t,y_i)$

$$L = \frac{f(t,y_2) - f(t,y_1)}{y_2 - y_1}$$

Entonces

$$L = \frac{\partial f}{\partial y}(t,y)$$

Si  $L \geq 0$ es máximo en intervalo $(y_1, y_2)$

$$\left| \frac{\partial f}{\partial y}(t,y) \right| \leq L$$

Quedando demostrado el teorema 1

#### **2.3. Definición 2 [1]**

El problema del valor inicial

$$\frac{dy}{dt} = f(t,y), \;\;\;\; a \leq t \leq b, \;\;\;\; y(a) = \alpha$$

se dice que es un problema bien planteado si:

* Existe una única solución $y(t)$, y

* Existen constantes $\epsilon_0 > 0$ y $k > 0$, tales que par cualquier $\epsilon$ en $(0,\epsilon_0)$, siempre que $\delta (t)$ es continua con $|\delta (t)| < \epsilon$ para toda $t$ en $[a,b]$, y cuando $|\delta_0| < \epsilon$, el problema de valor inicial

$$\frac{dz}{dt} = f(t,z) + \delta (t), \;\;\;\; a \leq t \leq b, \;\;\;\; z(a) = \alpha + \delta_0$$

tiene una única solución $z(t)$ que satisface

$$z(t) - y(t) < k\epsilon \;\;\;\;\; para\;toda\;t\;en\;[a,b].$$

#### **2.4. Teorema 2 [1]**

Suponga que $D = \{ (t,y) | a \leq t \leq b \; y \; -\infty < y < \infty \}$. Si $f$ es continua y satisface la condición de Lipschitz en la variable y sobre el conjunto $D$, entonces el problema de valor inicial

$$\frac{dy}{dt} = f(t,y), \;\;\;\; a \leq t \leq b, \;\;\;\; y(a) = \alpha$$

está bien planteado.

### **3. Práctica**

#### **3.1. Programas a utilizar**

In [1]:
#librerias necesarias
import math as mt
import numpy as np
import pandas as pd
import sys

#programas diseñados
ruta = 'C:\\Users\\Brayan Acosta\\Documents\\Semestre 2023-3\\Calculo Numerico\\Drive-Ric\\LAB-06'
sys.path.append(ruta)
from EDO import * #documentado en el archivo anexo

#### **3.2. Problema 1**

Aproximando la solución a $ y′ = (t + 2t^3)y^3 − ty, \;\; 0 \leq t \leq 2, \;\; y(0) = 1/3, \;\; h = 0,05$.

In [2]:
#definiendo parámetros
def f1(t, y):
    return (t + 2*y**3)*y**3 - t*y

a1, b1 = 0, 2
y01 = 1/3
h = 0.05
#recordando h = (b - a) / N
N1 = int((b1 - a1)/h)

#ecuación diferencial ordinaria
ed1 = EDO(a1, b1, y01, N1, f1)

##### **3.2.1. Método de Euler**

In [3]:
#resultado por el método de Euler
we1 = ed1.euler()

t0 = 0.00, w0 = 0.333333333333
t1  = 0.05, w1  = 0.333470507545
t2  = 0.10, w2  = 0.332867051489
t3  = 0.15, w3  = 0.331523152284
t4  = 0.20, w4  = 0.329442770369
t5  = 0.25, w5  = 0.326633739133
t6  = 0.30, w6  = 0.323107863936
t7  = 0.35, w7  = 0.318881011592
t8  = 0.40, w8  = 0.313973180183
t9  = 0.45, w9  = 0.308408538564
t10 = 0.50, w10 = 0.302215424924
t11 = 0.55, w11 = 0.295426294437
t12 = 0.60, w12 = 0.288077607277
t13 = 0.65, w13 = 0.280209650076
t14 = 0.70, w14 = 0.271866286139
t15 = 0.75, w15 = 0.263094632404
t16 = 0.80, w16 = 0.253944663983
t17 = 0.85, w17 = 0.244468750174
t18 = 0.90, w18 = 0.234721128820
t19 = 0.95, w19 = 0.224757328734
t20 = 1.00, w20 = 0.214633552446
t21 = 1.05, w21 = 0.204406033597
t22 = 1.10, w22 = 0.194130384791
t23 = 1.15, w23 = 0.183860952528
t24 = 1.20, w24 = 0.173650195901
t25 = 1.25, w25 = 0.163548104997
t26 = 1.30, w26 = 0.153601673498
t27 = 1.35, w27 = 0.143854437868
t28 = 1.40, w28 = 0.134346092845
t29 = 1.45, w29 = 0.125112189995
t30 = 1.50, 

Según el método de Euler:

$$y(2) = 0.047434199454$$

##### **3.2.2. Método de Runge-Kutta**

In [4]:
wrk1 = ed1.RK4()

t0 = 0.00, w0 = 0.333333333333
t1  = 0.05, w1  = 0.333100080123
t2  = 0.10, w2  = 0.332126310653
t3  = 0.15, w3  = 0.330415374075
t4  = 0.20, w4  = 0.327974479195
t5  = 0.25, w5  = 0.324814782141
t6  = 0.30, w6  = 0.320951463727
t7  = 0.35, w7  = 0.316403786675
t8  = 0.40, w8  = 0.311195122423
t9  = 0.45, w9  = 0.305352937436
t10 = 0.50, w10 = 0.298908729728
t11 = 0.55, w11 = 0.291897907662
t12 = 0.60, w12 = 0.284359604946
t13 = 0.65, w13 = 0.276336428034
t14 = 0.70, w14 = 0.267874134708
t15 = 0.75, w15 = 0.259021245385
t16 = 0.80, w16 = 0.249828591550
t17 = 0.85, w17 = 0.240348808427
t18 = 0.90, w18 = 0.230635781540
t19 = 0.95, w19 = 0.220744059033
t20 = 1.00, w20 = 0.210728243279
t21 = 1.05, w21 = 0.200642376508
t22 = 1.10, w22 = 0.190539335671
t23 = 1.15, w23 = 0.180470251582
t24 = 1.20, w24 = 0.170483966566
t25 = 1.25, w25 = 0.160626543320
t26 = 1.30, w26 = 0.150940835745
t27 = 1.35, w27 = 0.141466130021
t28 = 1.40, w28 = 0.132237861567
t29 = 1.45, w29 = 0.123287410721
t30 = 1.50, 

Según el método de Runge-Kutta:

$$y(2) = 0.048051751405$$

##### **3.2.3. Comprobación con libreria sympy**

In [31]:
import sympy as sp
from sympy.abc import x,z,t

C1 = sp.Symbol('C1')
y = sp.Function('y')

#construyendo ecuacion diferencial
f1sp = (t + 2*t**3)*y(t)**3 - t*y(t)
eq = sp.Eq(y(t).diff(t), f1sp)
eq


Eq(Derivative(y(t), t), -t*y(t) + (2*t**3 + t)*y(t)**3)

In [33]:
# obteniendo familia de soluciones
sol = sp.dsolve(y(t).diff(t) - f1sp)[1]
sol

Eq(y(t), sqrt(1/(C1*exp(t**2) + 2*t**2 + 3)))

In [34]:
#obteniendo solución
ics = {y(0) : 1/3}
Ceq = sp.Eq(sol.lhs.subs(t, 0).subs(ics), sol.rhs.subs(t, 0))
C = sp.solve(Ceq)[0]
solt = sol.subs(C1, C)
solt

Eq(y(t), sqrt(1/(2*t**2 + 6.0*exp(t**2) + 3)))

In [35]:
#evaluando en el punto de interés
solt.subs(t, 2).evalf()

Eq(y(2), 0.0543455066126645)

#### **3.3. Problema 2**

Aproximando la solución al sistema,

$$\left\{ \begin{array}{rlcc}
    u_1' = & u_2, & 0 \leq t \leq 2, & u_1(0) = 1; \\
    u_2' = & −u_1 − 2e^t + 1, & 0 \leq t \leq 2, & u_2(0) = 0; \\
    u_3' = & −u_1 − e^t + 1, & 0 \leq t \leq 2, & u_3(0) = 1;
\end{array} \right.$$

Con $h = 0.05$

In [5]:
def f2(t, vec):
    u1, u2, u3 = vec[0], vec[1], vec[2]
    du1 = u2
    du2 = -u1 -2*mt.exp(t) +1
    du3 = -u1 -mt.exp(t) +1
    return np.array([du1, du2, du3])

u10, u20, u30 = 1, 0, 1
vec = [u10, u20, u30]

ed2 = EDO(0, 2, vec, N1, f2)

##### **3.3.1. Método de Euler**

In [6]:

we2 = ed2.sis_euler()

t_0 = 0.00, u1_0 = 1.000000000000, u2_0 = 0.000000000000, u3_0 = 1.000000000000
t1  = 0.05, u1_1  = 1.000000000000, u2_1  = -0.100000000000, u3_1  = 0.950000000000
t2  = 0.10, u1_2  = 0.995000000000, u2_2  = -0.205127109638, u3_2  = 0.897436445181
t3  = 0.15, u1_3  = 0.984743644518, u2_3  = -0.315394201445, u3_3  = 0.842427899277
t4  = 0.20, u1_4  = 0.968973934446, u2_4  = -0.430814807944, u3_4  = 0.785099004915
t5  = 0.25, u1_5  = 0.947433194049, u2_5  = -0.551403780482, u3_5  = 0.725580170285
t6  = 0.30, u1_6  = 0.919863005025, u2_6  = -0.677177981853, u3_6  = 0.664007239748
t7  = 0.35, u1_7  = 0.886004105932, u2_7  = -0.808157012862, u3_7  = 0.600521149118
t8  = 0.40, u1_8  = 0.845596255289, u2_8  = -0.944363973018, u3_8  = 0.535267566392
t9  = 0.45, u1_9  = 0.798378056638, u2_9  = -1.085826255547, u3_9  = 0.468396518745
t10 = 0.50, u1_10 = 0.744086743861, u2_10 = -1.232576376928, u3_10 = 0.400062006639
t11 = 0.55, u1_11 = 0.682457925014, u2_11 = -1.384652841191, u3_11 = 0.330421605

Según el método de Euler:

$$\begin{split}
    u_1(2) & = -5.672240001303 \\
    u_2(2) & = -8.755031476292 \\ 
    u_3(2) & = -1.524370781938 \\  
\end{split}$$

##### **3.3.2. Método de Runge-Kutta**

In [7]:
wrk2 = ed2.sis_RK4()

t_0 = 0.00, u1_0 = 1.000000000000, u2_0 = 0.000000000000, u3_0 = 1.000000000000
t1  = 0.05, u1_1  = 0.997458328966, u2_1  = -0.102499998910, u3_1  = 0.948771097578
t2  = 0.10, u1_2  = 0.989666655569, u2_2  = -0.210000156236, u3_2  = 0.895170762068
t3  = 0.15, u1_3  = 0.976374955840, u2_3  = -0.322501276682, u3_3  = 0.839332966398
t4  = 0.20, u1_4  = 0.957333135494, u2_4  = -0.440005482664, u3_4  = 0.781397275977
t5  = 0.25, u1_5  = 0.932290946600, u2_5  = -0.562516917443, u3_5  = 0.721508499861
t6  = 0.30, u1_6  = 0.900997868305, u2_6  = -0.690042479530, u3_6  = 0.659816328805
t7  = 0.35, u1_7  = 0.863202950065, u2_7  = -0.822592588394, u3_7  = 0.596474961109
t8  = 0.40, u1_8  = 0.818654615795, u2_8  = -0.960181981500, u3_8  = 0.531642717208
t9  = 0.45, u1_9  = 0.767100427394, u2_9  = -1.102830542771, u3_9  = 0.465481643953
t10 = 0.50, u1_10 = 0.708286806053, u2_10 = -1.250564162534, u3_10 = 0.398157109573
t11 = 0.55, u1_11 = 0.641958709774, u2_11 = -1.403415629109, u3_11 = 0.329837390

Según el método de Runge-Kutta:

$$\begin{split}
    u_1(2) & = -5.895905281674 \\
    u_2(2) & = -8.714499869831 \\ 
    u_3(2) & = -1.325443757036 \\  
\end{split}$$

#### **4. Análisis de resultados**

En el problema N°1 realizando una comparación del método de Euler y el de Runge-Kutta para la resolución de una ecuación diferencial ordinaria, se obtuvieron resultados aproximadamente iguales en el orden decimal de los $10^{-3}$ para la función evaluada $y(t_0)$ con $t_0$ dentro del intervalo dado. Cabe señalar el error del resultado respecto a la solución hallada con la libreria sympy de python con respecto al método de Runge-Kutta $|y_{sympy}(2) - y_{RK4}(2)| = 0.054345506613 - 0.048051751405 = 0.006293755208$

Para el problema N°2 también se realizó una comparación del método de Euler y el de Runge-Kutta para la resolución de un sistema de ecuaciones diferenciales ordinarias, donde las variables dependientes corresponden a $u_1$, $u_2$ y $u_3$. Por consiguiente, se obtuvieron resultados aproximadamente iguales en el orden decimal de los $10^{-1}$ para la función evaluada $u_i(t_0), \; \forall \; i = 1,2,3$ con $t_0$ dentro del intervalo dado.


#### **5. Conclusión**

Debido al proceso numérico que se realizó en el programa EDO y observando los resultados obtenidos y el menor error, se puede deducir que el método de Runge-Kutta es una mejor aproximación a la solución deseada respecto al método de Euler, sin embargo para grandes intervalos con divisiones en el orden de los millones se puede obterner una estimación por el método de Euler, ya que utiliza procesos más rápidos. Por otra parte hay que considerar que a medida que el punto a evaluar se aleja de la condición inicial la aproximación va alejandose considerablemente del resultado deseado.

Los Métodos analizados anteriormente son factibles a la hora de resolver un sistema de ecuaciones diferenciales ordinarias, no obstante, es recomendable utilizar notaciones vectoriales para estos casos.

#### **6. Referencias**

[1] **Analisis Numérico**, *Richard L. Burden • Douglas J. Faires • Annette M. Burden*, 10ma edición

[2] **Métodos Numéricos con Python**, *Ovalle D., Bernal M., Posada J.*, Editorial Politecnico Grancolombiano, (2021).
